Framework/MDAlgorithms/src/CompareMDWorkspaces.cpp

```diff
@@ -210,28 +363,83 @@ void CompareMDWorkspaces::compareMDWorkspaces(typename MDEventWorkspace<MDE, nd>
     auto *gridbox1 = dynamic_cast<MDGridBox<MDE, nd> *>(box1);
     auto *gridbox2 = dynamic_cast<MDGridBox<MDE, nd> *>(box2);
     if (gridbox1 && gridbox2) {
+      // MDGridBox: compare box size on each dimension
       for (size_t d = 0; d < nd; d++)
         this->compareTol(gridbox1->getBoxSize(d), gridbox2->getBoxSize(d), "Box sizes do not match");
-    }
+    } else {
+      // Could be both MDBoxes (with events)
+      auto *mdbox1 = dynamic_cast<MDBox<MDE, nd> *>(box1);
+      auto *mdbox2 = dynamic_cast<MDBox<MDE, nd> *>(box2);
+      g_log.debug() << "Box " << ibox << "ws1 npoints = " << box1->getNPoints()
+                    << "; ws2 npoints = " << box2->getNPoints() << "\n";
+
+      // Rule out the case if one and only one box is MDBox
+      if (mdbox1 && !mdbox2) {
+        // workspace 2's is MDBox but workspace 1 is not
+        throw CompareFailsException("Worksapce 2's Box " + std::to_string(ibox) + " is not MDBox");
+      } else if (!mdbox1 && mdbox2) {
+        // workspace 2's is MDBox but workspace 1 is not
+        throw CompareFailsException("Worksapce 1's Box " + std::to_string(ibox) + " is not MDBox");
+      }
 
-    // Are both MDBoxes (with events)
-    auto *mdbox1 = dynamic_cast<MDBox<MDE, nd> *>(box1);
-    auto *mdbox2 = dynamic_cast<MDBox<MDE, nd> *>(box2);
-    if (mdbox1 && mdbox2) {
+      // Both boxes are MDBoxes:
       if (m_CheckEvents) {
         const std::vector<MDE> &events1 = mdbox1->getConstEvents();
         const std::vector<MDE> &events2 = mdbox2->getConstEvents();
+
         try {
           this->compare(events1.size(), events2.size(), "Box event vectors are not the same length");
+
           if (events1.size() == events2.size() && events1.size() > 2) {
-            // Check first and last event
+
+            std::vector<SimpleMDEvent> events_vec1;
+            std::vector<SimpleMDEvent> events_vec2;
+
+            // convert MDEvents vectors to SimpleMDEvent vectors for comparison
             for (size_t i = 0; i < events1.size(); i++) {
+              std::vector<float> centers1;
```

Comment:
I think it runs faster if youallocate `std::vector<float> centers1(nd)` and then assign `centers1[d] = events1[i].getCenter(d)`

Reply:
Good point.  I will make test to see the difference.